### PR TITLE
[SPARK-42518] [CONNECT] Scala Client DataFrameWriterV2

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?
Adding DataFrameWriterV2. This allows users to use the Dataset#writeTo API.

### Why are the changes needed?
Impls Dataset#writeTo

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
E2E

This is based on https://github.com/apache/spark/pull/40061

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 1]
```diff
@@ -156,6 +156,19 @@ class ClientE2ETestSuite extends RemoteSparkSession {
     }
   }
 
+  test("write v2") {
+    try {
+      spark.range(3).writeTo("myTableV2").using("parquet").create()
```
Comment (zhenlineo) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:161]: This API feels wrong, a few errors I get with this API:
```
spark.range(3).writeTo("myTableV2").create() -> Error: should specify source
```

```
spark.range(3).writeTo("myTableV2").using("parquet").createOrReplace() -> Error: dose not support create or select.
```

```
spark.range(3).writeTo("myTableV2").using("parquet").create()
spark.range(3).writeTo("myTableV2").append() -> Error: Cannot write into v1 table
```

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:161]: hmm... when I add `parquet-hadoop` as a test dependency of `connect-client-jvm` module, the test passed ....

**Comment by LuciferYang** [comment]
Move `SaveMode` to catalyst module is a break change, need add `ProblemFilters` to `MimaExcludes`

**Comment by amaliujia** [comment]
> Move `SaveMode` to catalyst module is a break change, need add `ProblemFilters` to `MimaExcludes`
I think we shouldn't make such breaking change?

**Comment by LuciferYang** [comment]
> > Move `SaveMode` to catalyst module is a break change, need add `ProblemFilters` to `MimaExcludes`
> 
> I think we shouldn't make such breaking change?

best to avoid

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:161]: One problem is that I'm not sure why the test dependency `parquet-hadoop` need to be added to the `connect-client-jvm `

**Comment by zhenlineo** [comment]
Depends on https://github.com/apache/spark/pull/40109 to get the correct error message for `create`.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 1 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:162]: After https://github.com/apache/spark/pull/40109, the current behavior of this client is the same as pyspark.

Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:162]: For my understanding, the diff between client test and [server test](https://github.com/apache/spark/blob/master/connector/connect/server/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala#L681-L706), it that the config setting:
```
spark.conf.set("spark.sql.catalog.testcat", classOf[InMemoryTableCatalog].getName)
```
changes the current catalog to a V2 catalog.

@cloud-fan Could you explain the diff better for me? I saw you wrote the [SQL test](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DataFrameWriterV2Suite.scala#L55) which also have this settings?

CC @hvanhovell

**Comment by hvanhovell** [comment]
merging.

**Pull request closed without merging**